### DISCUSSION TITLE
Is that possible to ignore ForeignKeyConstraint for relationship joins?

### DISCUSSION THREAD

**Discussion by AlexanderFarkas** [sqlalchemy/sqlalchemy#10723] [Usage Questions] [answered]
I have multitenant single-schema approach.
Each entity contains "tenant_id" column, and also each entity has ForeignKeyConstraint enforcing the same "tenant_id" for each of the relations.

So "order" has relationship to payment:
```python
Column("tenant_id", ForeignKey("tenant.id"))
relationship("Payment")
# ....
ForeignKeyConstraint(["payment_id", "tenant_id"], ["payment.id", "payment.tenant_id"])
```

But SQLAlchemy adds both "payment.id" and "payment.company_id" to "synchronize_pairs" in relationship, so setting `order.payment` to `None` makes `order.tenant_id` None also, which is not desired.

The possible solution is specify "foreign_keys" explicitly for each of the relationships, so it ignores `tenant_id` entirely, but then I have to do it on every "multitenant" entity.

The other hypothetical solution is to add direct foreign key to "payment_id" and somehow make SQLAlchemy ignore the composite foreign key.

Maybe there is a better approach.

**Comment by CaselIT** [comment]
Hi,

All this is documented here https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship
Specifically https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.foreign_keys

**Reply by AlexanderFarkas** [reply]
But if I specify `foreign_keys` to only `payment_id`, join will happen on only payment_id, and it won't use the database index.

**Reply by AlexanderFarkas** [reply]
Am I correct, that in my case I should specify both primary_join (on composite key) and `foreign_keys=["payment_id"]`

**Reply by CaselIT** [reply]
I may have misunderstood what you are trying to do, sorry.
I'm not sure if it's possible to configure `synchronize_pairs` manually. @zzzeek do you have some suggestions for this use case?

**Reply by AlexanderFarkas** [reply]
Actually, I found pretty much the same [situation](https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#overlapping-foreign-keys). Documentation is golden, thank you guys for your hard work.

**Comment by zzzeek** [comment, marked as answer]
> 
> The possible solution is specify "foreign_keys" explicitly for each of the relationships, so it ignores `tenant_id` entirely, 

This would be the correct approach

> but then I have to do it on every "multitenant" entity.

what's the issue there?  every entity has ForeignKey directives, relationship directives, etc., that's all on every one of them also...if you have a standard pattern you can always make a Python function that returns relationship() with the parameters you want
